x/profile/internal/keeper/keeper.go

```diff
@@ -0,0 +1,132 @@
+	"github.com/desmos-labs/desmos/x/profile/internal/types"
+)
+
+// Keeper maintains the link to data storage and exposes getter/setter methods for the various parts of the state machine
+type Keeper struct {
+	StoreKey sdk.StoreKey // Unexposed key to access store from sdk.Context
+	Cdc      *codec.Codec // The wire codec for binary encoding/decoding.
+}
+
+// NewKeeper creates new instances of the magpie Keeper
+func NewKeeper(cdc *codec.Codec, storeKey sdk.StoreKey) Keeper {
+	return Keeper{
+		StoreKey: storeKey,
+		Cdc:      cdc,
+	}
+}
+
+// AssociateMonikerWithAddress save the relation of moniker and address on chain
+func (k Keeper) AssociateMonikerWithAddress(ctx sdk.Context, moniker string, address sdk.AccAddress) {
+	store := ctx.KVStore(k.StoreKey)
+	key := types.MonikerStoreKey(moniker)
+	store.Set(key, k.Cdc.MustMarshalBinaryBare(&address))
+}
+
+// GetMonikerRelatedAddress returns the address associated to the given moniker or nil if it not exists
+func (k Keeper) GetMonikerRelatedAddress(ctx sdk.Context, moniker string) (addr sdk.AccAddress) {
+	store := ctx.KVStore(k.StoreKey)
+	bz := store.Get(types.MonikerStoreKey(moniker))
+	if bz == nil {
+		return nil
+	}
+	k.Cdc.MustUnmarshalBinaryBare(bz, &addr)
+	return addr
+}
+
+// GetMonikerFromAddress returns the moniker associated with the given address or an empty string if no moniker exists
+func (k Keeper) GetMonikerFromAddress(ctx sdk.Context, addr sdk.AccAddress) (moniker string) {
+	store := ctx.KVStore(k.StoreKey)
+	it := sdk.KVStorePrefixIterator(store, types.MonikerStorePrefix)
+	defer it.Close()
+
+	for ; it.Valid(); it.Next() {
+		var acc sdk.AccAddress
+		k.Cdc.MustUnmarshalBinaryBare(it.Value(), &acc)
+		if acc.Equals(addr) {
+			return string(bytes.TrimPrefix(it.Key(), types.MonikerStorePrefix))
+		}
+	}
+
+	return ""
+}
+
+// DeleteMonikerAddressAssociation delete the given moniker association with an address
+func (k Keeper) DeleteMonikerAddressAssociation(ctx sdk.Context, moniker string) {
+	store := ctx.KVStore(k.StoreKey)
+	key := types.MonikerStoreKey(moniker)
+	store.Delete(key)
+}
+
+// SaveProfile allows to save the given account inside the current context.
+// It assumes that the given account has already been validated.
+// It returns an error if an account with the same moniker from a different creator already exists
+func (k Keeper) SaveProfile(ctx sdk.Context, profile types.Profile) error {
+	store := ctx.KVStore(k.StoreKey)
+
+	key := types.ProfileStoreKey(profile.Creator.String())
+
+	moniker := k.GetMonikerFromAddress(ctx, profile.Creator)
+	k.DeleteMonikerAddressAssociation(ctx, moniker)
+
+	if addr := k.GetMonikerRelatedAddress(ctx, profile.Moniker); addr != nil && !addr.Equals(profile.Creator) {
+		return fmt.Errorf("an account with moniker: %s has already been created", profile.Moniker)
+	}
+
+	k.AssociateMonikerWithAddress(ctx, profile.Moniker, profile.Creator)
+	store.Set(key, k.Cdc.MustMarshalBinaryBare(&profile))
+
+	return nil
```

Comment:
This method looks a little bit clumsy. Maybe we can improve how it works by refactoring it a bit. 

First, let's see what it should do: 

1. Get any current profile with the same moniker. 
2. Check for the ownership of such profile. 
3. If the existing profile is owned by the owner of the new one, edit it. Otherwise throw an error. 

So, I think this might be changed as follow: 

```golang
	existingProfile := k.GetProfileByMoniker(profile.Moniker)
	if !existingProfile.Creator.Equals(profile.Creator) {
		return fmt.ErrorF("...")
	}

	k.replaceMoniker(ctx, existingProfile.Moniker, profile.Moniker)

	store := ctx.KVStore(k.StoreKey)
	store.Set(types.ProfileStoreKey(profile.Creator), k.Cdc.MustMarshalBinaryBare(&profile))
	return nil
```

With `replaceMoniker` being (note that this is private so that it is only used inside the keeper): 

```golang
func (k Keeper) replaceMoniker(ctx sdk.Context, oldMoniker, newMoniker string) {
	// code of k.DeleteMonikerAddressAssociation(ctx, existingProfile.Moniker)
	// code of k.AssociateMonikerWithAddress(ctx, profile.Moniker, profile.Creator)
}
```